### ISSUE TITLE
Update dice function

### ISSUE THREAD

**Issue by IliaLarchenko** [catalyst-team/catalyst#522]
I think that the current version of `dice` is not 100% correct (but maybe there is some rationale behind it I don't see).

Current approach:
```
intersection = torch.sum(targets * outputs)
union = torch.sum(targets) + torch.sum(outputs)
dice = 2 * intersection / (union + eps)
```

1. In the case of empty targets, mask and empty predicted mask `intersection == 0` therefore `dice == 0`. But it seems wrong because everything is predicted correctly. The small change can fix it:
```
intersection = torch.sum(targets * outputs)
union = torch.sum(targets) + torch.sum(outputs)
dice = (2 * intersection + eps) / (union + eps)
```

2. The metric is computed as if `targets` and `outputs` were the ground truth and predicted values of the one single-channel mask. But this formula is not right if there are several channels in the mask or the batch_size > 1. Even in the case of 2 samples in batch or 2-channel mask (with possible intersections), the formula is not correct.

```
2 * (intersection1 + intersection2) / (union1 + union1) != intersection1/union1 + intersection2/union2
```

In the case of several channels and large batches, the difference can be more significant. I think the metric should be computed on each channel of each sample separately and only then averaged.

If we work with batches of shape [N x H x W x CH] it can be solved like this:
```
intersection = (targets * outputs).mean(dim=2).mean(dim=2)
union = targets.mean(dim=2).mean(dim=2) + outputs.mean(dim=2).mean(dim=2)
dice = ((2 * intersection + eps) / (union + eps)).mean()
``` 
As I've done here: https://github.com/IliaLarchenko/catalyst/blob/master/catalyst/dl/utils/criterion/dice.py

But I'm not sure if there are other use cases with different data shapes where this approach will not work.

**Comment by Scitator** [comment]
@pokidyshev @TezRomacH  Have not we already solve this issue?

**Comment by Scitator** [comment]
looks like done
